rfc-119-merge-mainstream-browse-topics-and-taxons.md

```diff
@@ -0,0 +1,169 @@
+
+Generally there are two main type of information we want to tranfer from mainstream browse and specialist topic to the equivalent taxon. The first is the links to content items. Currently content items (guidance pages, manuals etc..) do not contain links to mainstream browse or specialist topics. It's the reverse, mainstream browse pages and specialist topics have a list of content that's associated to them. This list we can migrate over to the taxon.
+
+Secondly, we'd carry across the relationships (links) between mainstream browse pages/specialist topics. This manifests as multiple hierarchies within the same taxons. This is because different relationships are useful for different things ie. the currently hierarchy for taxonomy is good for emails subscriptions and organising taxons, but not so great at modelling navigation. So we'd bring over the mainstream browse hierarchy to serve that purpose.
+
+#### Top level attributes
+
+All top level attributes will continue to have the existing values in the taxon:
+
+- `title`
+- `description`
+- `analytics_identifier`
+- `content_id`
+- `document_type`
+- `first_published_at`
+- `locale`
+- `phase`
+- `public_updated_at`
+- `publishing_scheduled_at`
+- `rendering_app`
+- `scheduled_publishing_delay_seconds`
+- `schema_name`
+- `updated_at`
+- `withdrawn_notice`
+- `publishing_request_id`
+
+This is because they don't cause conflict or the mainstream or specialist topic value becomes redundant.
+
+- `publishing_app`
+
+Currently references Collections Publisher for mainstream and specialist topics, however references Content Tagger for taxons. We propose that `publishing-app` remains Content Tagger as it still has the responsibility of creating and removing taxons, even though some sub properties (i.e. links and details) may be set by Collections Publisher. In the future, we could also look at consolidating the publishing functionality of Collections Publisher and Content Tagger.
+
+- `base_path`
+
+Mainstream browse and specialist topic pages have different `base_path` to taxons. We propose that taxons keep their existing `base_path`. We could use new mainstream browse/specialist topic links within taxons to generate old base paths or find the appropriate taxon from an existing mainstream browse or specialist topic base path. Alternatively, we keep the old base paths as renamed attributes (e.g. `mainstream_browse_base_path`).
+
+#### Link attributes
+
+Taxons continue to have the following links:
+- `legacy_taxons` (potentially remove once specialist topics are removed)
+- `parent_taxons`
+- `available_translations`
+
+Add from mainstream browse:
+- `browse_parent` (parent taxons which should be used for mainstream browse)
+- `browse_children` (reverse links to `browse_parent`)
+
+Add from specialist topic:
+- `specialist_topic_parent` (parent taxons which should be used for specialist topics)
```

Comment:
I can sort of see the need for a taxon to have "specialist children". How come it needs a different parent link than `parent_taxon`, though?

Reply:
This is because the organisational hierarchy of specialist topics is different than taxons. 

For example, the specialist topic parent for 'motorways-major-roads' is 'transport', but in the taxonomy the parent for 'motorways-major-roads' might be ' driving-and-road-transport' (who's parent is 'transport'). This is needed to render pages like (https://www.gov.uk/topic/transport/motorways-major-roads)

Specialist topic links:
'motorways-major-roads' -> 'transport'

Taxon links:
'motorways-major-roads' -> 'driving-and-road-transport' -> 'transport'

Does that make sense?

Reply:
Oh, I see, yes. I guess it's kind of obvious: if there are different children, there will be different parents 🙄.

Now I appreciate the hierarchical distinction, I'm starting to worry about the end result, which could get quite confusing: we really are modelling three hierarchies in one object. I can see this leading to synchronisation issues where, for example, a taxon is moved to new parent, but it's mainstream browse and specialist topic hierarchies are left as-is. We could end up with quite a complex UI to manage the different levels of parents/children.

Since the aim of this RFC is to align these three things, I'm wondering if we could go further: could we just have one hierarchy (of taxons), and use flags to filter the hierarchy into a mainstream browse hierarchy (for navigation), and a specialist topic hierarchy (still not sure what this one is even for 😆)?

So, in your example, we could have:

'motorways-major-roads' (`browse: true`) -> 'driving-and-road-transport' (`browse: false`) -> 'transport' (`browse: true`)

I guess a disadvantage of this is that the rendering apps would have a harder time using the hierarchy, since they'd have to manually skip between levels. However, that seems reasonable, since the concern seems to be primarily a rendering one, and we need to load [the whole hierarchy](https://www.gov.uk/browse) then anyway.

Even if we don't go down this route, it would be helpful to explain why it's a rejected approach. Sorry, I know that's a fair chunk of effort. Let's chat if you want to go over anything!

Reply:
Thanks @benthorner! I'm interested in your comment

 > I can see this leading to synchronisation issues where, for example, a taxon is moved to new parent, but it's mainstream browse and specialist topic hierarchies are left as-is. 

So your example where changing a taxon parent would't effect the mainstream browse or specialist topic parents. I think this is desired, the taxon hierarchy isn't used for navigation. Having separate hierarchies prevents side effects e.g. we don't want to change the browse pages because we changed the taxon hierarchy for another reason. Can you think of any specific examples where this an issue?

> We could end up with quite a complex UI to manage the different levels of parents/children.

So currently, we already have multiple UIs and complexity in the publishing apps to deal with the three different categorisation types (mainstream browse, topic, taxons). Do you think the complexity is going to worse?

Thanks also for you suggestion about an alternative approach:

> use flags to filter the hierarchy into a mainstream browse hierarchy (for navigation), and a specialist topic hierarchy (still not sure what this one is even for 😆)?

So I think this assumes that siblings in mainstream browse hierarchy share the same parent in the taxon hierarchy - which isn't true. For example:

![image](https://user-images.githubusercontent.com/11051676/99429551-1391e780-2900-11eb-9c26-962825509486.png)

Reply:
Oh wow, nice diagram! The criss-cross hierarchy is definitely a surprise, but if it's a product need, then I agree we need to support it and the additional links make sense. I do think we should explain more about the hierarchies in the background section, though, perhaps with a version of this diagram to support it. I appreciate that's a fair amount of effort. What I'm finding with this RFC is that there's actually quite a lot of hidden knowledge with these hierarchies, which makes it hard to understand the proposal.

The need for separate hierarchies does actually lead me to question the value we get from merging these three things: if the hierarchies are meant to deviate, then would it be clearer to continue representing them as three distinct things, despite some commonalities they may have? What problems are we solving by combining them? I can see there might be some value in not having to synchronise titles, for example. Could we also elaborate the problem section to explain more specifically what we're concerned about with divergence?